src/drivers/virtio/virtio-pci.zig

```diff
@@ -0,0 +1,309 @@
+// Descriptor iterator helper.
+pub const DescIter = struct {
+  drv: *Driver,
+  i: u8,
+  curr: Descriptor,
+  next: Descriptor,
+  // Put the head of a descriptor chain on the available ring
+  pub fn begin(iter: *DescIter) void {
+    var i = &iter.drv.queues[iter.i];
+    iter.next = iter.drv.descr(iter.i);
+    i.avail.rings((i.avail.idx + i.pending) % i.size).* = iter.next;
+    i.pending += 1;
+  }
+  // Put a descriptor to be part of the descriptor chain
+  pub fn put(iter: *DescIter, a: anytype, len: u32, flags: u16) void {
+    iter.curr = iter.next;
+    iter.next = if ((flags & VRING_DESC_F_NEXT) != 0) iter.drv.descr(iter.i) else 0xFFFF;
+    assert(len <= 0x1000);
+    const addr = paging.translate_virt(@ptrToInt(a), null) catch |err| {
+      @panic("virtio-pci: can't get the physical address");
+    };
+    iter.drv.queues[iter.i].desc[iter.curr] = .{ .addr = addr, .len = len, .flags = flags, .next = iter.next };
+  }
+};
+
+// The actual driver structure for the modern virtio-pci transport. It expects already-initialized BARs.
```

Comment:
In the future, use `///` for doc comments. Tools like zls will display them automatically